File: src/utils/log.ts
// Copyright (c) shigeru22. Licensed under the MIT license.
// See LICENSE in the repository root for details.

import chalk from "chalk";
import { DateUtils } from "./date.js";
import { Environment, NodeEnvironment } from "./env.js";

export enum LogSeverity {
  CRITICAL,
  ERROR,
  WARNING,
  INFO,
  VERBOSE,
  DEBUG
}

const severityCode = ["C", "E", "W", "I", "V", "D"];

export class Log {
  public static write(severity: LogSeverity, source: string, message: string) {
    const logText = this.createLogText(severity, source, message);

    switch(severity) {
      case LogSeverity.CRITICAL: // fallthrough
      case LogSeverity.ERROR:
        console.error(chalk.red(logText));
        break;
      case LogSeverity.WARNING:
        console.warn(chalk.yellow(logText));
        break;
      case LogSeverity.INFO:
        console.log(logText);
        break;
      case LogSeverity.VERBOSE: // fallthrough
      case LogSeverity.DEBUG:
        console.debug(chalk.gray(logText));
        break;
    }
  }

  public static writeCritical(source: string, message: string) {
    if(Environment.getLogSeverity() >= LogSeverity.CRITICAL) {
      this.write(LogSeverity.CRITICAL, source, message);
    }
  }

  public static writeError(source: string, message: string) {
    if(Environment.getLogSeverity() >= LogSeverity.ERROR) {
      this.write(LogSeverity.ERROR, source, message);
    }
  }

  public static writeWarning(source: string, message: string) {
    if(Environment.getLogSeverity() >= LogSeverity.WARNING) {
      this.write(LogSeverity.WARNING, source, message);
    }
  }

  public static writeInfo(source: string, message: string) {
    if(Environment.getLogSeverity() >= LogSeverity.INFO) {
      this.write(LogSeverity.INFO, source, message);
    }
  }

  public static writeVerbose(source: string, message: string) {
    if(Environment.getEnvironment() <= NodeEnvironment.STAGING && Environment.getLogSeverity() >= LogSeverity.VERBOSE) {
      this.write(LogSeverity.VERBOSE, source, message);
    }
  }

  public static writeDebug(source: string, message: string) {
    if(Environment.getEnvironment() <= NodeEnvironment.DEVELOPMENT && Environment.getLogSeverity() >= LogSeverity.DEBUG) {
      this.write(LogSeverity.DEBUG, source, message);
    }
  }

  public static createLogText(severity: LogSeverity, source: string, message: string): string {
    // eslint-disable-next-line operator-linebreak
    return `${ DateUtils.getCurrentDateTimeString(Environment.getShouldUseUtc()) } :: ` +
      `${ severityCode[severity] } :: ${ source } :: ${ message }`;
  }

  public static createExceptionMessage(err: Error, errorMessage: string): string {
    return `${ errorMessage }\n${ err.stack === undefined ? `${ err.name }: ${ err.message }` : err.stack }`;
  }
}

File: SadBot.Utils/Log.cs
// Copyright (c) shigeru22. Licensed under the MIT license.
// See LICENSE in the repository root for details.

using System.Diagnostics;
using System.Reflection;
using System.Runtime.CompilerServices;

namespace SadBot.Utils;

public enum LogSeverity
{
	Critical = 0,
	Error,
	Warning,
	Info,
	Verbose,
	Debug
}

public static class Log
{
	private static readonly string[] logSeverityString = {
		"CRITICAL",
		"ERROR",
		"WARNING",
		"INFO",
		"VERBOSE",
		"DEBUG"
	};

	public static void Write(LogSeverity severity, string message, [CallerMemberName] string source = "")
	{
		switch (severity)
		{
			case LogSeverity.Critical:
				WriteCritical(message, source);
				break;
			case LogSeverity.Error:
				WriteError(message, source);
				break;
			case LogSeverity.Warning:
				WriteWarning(message, source);
				break;
			case LogSeverity.Info:
				WriteInfo(message, source);
				break;
			case LogSeverity.Verbose:
				WriteVerbose(message, source);
				break;
			case LogSeverity.Debug:
				WriteDebug(message, source);
				break;
			default:
				throw new ArgumentOutOfRangeException(nameof(severity), severity, null);
		}
	}

	public static Task WriteAsync(LogSeverity severity, string message, [CallerMemberName] string source = "")
	{
		Write(severity, message, source);
		return Task.CompletedTask;
	}

	public static void WriteCritical(string message, [CallerMemberName] string source = "")
	{
		ConsoleColor currentColor = Console.ForegroundColor;
		Console.ForegroundColor = ConsoleColor.Red;

		Console.WriteLine(GenerateLogMessage(LogSeverity.Critical, message, source));

		Console.ForegroundColor = currentColor;
	}

	public static void WriteError(string message, [CallerMemberName] string source = "")
	{
		ConsoleColor currentColor = Console.ForegroundColor;
		Console.ForegroundColor = ConsoleColor.Red;

		Console.WriteLine(GenerateLogMessage(LogSeverity.Error, message, source));

		Console.ForegroundColor = currentColor;
	}

	public static void WriteWarning(string message, [CallerMemberName] string source = "")
	{
		ConsoleColor currentColor = Console.ForegroundColor;
		Console.ForegroundColor = ConsoleColor.Yellow;

		Console.WriteLine(GenerateLogMessage(LogSeverity.Warning, message, source));

		Console.ForegroundColor = currentColor;
	}

	public static void WriteInfo(string message, [CallerMemberName] string source = "")
	{
		Console.WriteLine(GenerateLogMessage(LogSeverity.Info, message, source));
	}

	public static void WriteVerbose(string message, [CallerMemberName] string source = "")
	{
		ConsoleColor currentColor = Console.ForegroundColor;
		Console.ForegroundColor = ConsoleColor.DarkGray;

		Console.WriteLine(GenerateLogMessage(LogSeverity.Verbose, message, source));

		Console.ForegroundColor = currentColor;
	}

	public static void WriteDebug(string message, [CallerMemberName] string source = "")
	{
		ConsoleColor currentColor = Console.ForegroundColor;
		Console.ForegroundColor = ConsoleColor.DarkGray;

		Console.WriteLine(GenerateLogMessage(LogSeverity.Debug, message, source));

		Console.ForegroundColor = currentColor;
	}

	public static void DeletePreviousLine(bool keepCurrentLine = false)
	{
		int currentCursorLine = Console.CursorTop;
		Console.SetCursorPosition(0, currentCursorLine - 1);
		Console.Write(new string(' ', Console.WindowWidth));

		if (!keepCurrentLine)
		{
			Console.SetCursorPosition(0, currentCursorLine - 1);
		}
	}

	public static string GenerateLogMessage(LogSeverity severity, string message, string source)
	{
		string tempSource = source.Equals(".ctor") || source.Equals(".cctor") ? GetParentTypeName() : source;
		return $"{Date.GetCurrentDateTime()} :: {logSeverityString[(int)severity][0]} :: {tempSource} :: {message}";
	}

	public static string GenerateExceptionMessage(Exception e, string errorMessage)
	{
		return $"{errorMessage}. Exception details below.\n{e}";
	}

	private static string GetParentTypeName()
	{
		StackFrame stackFrame = new StackFrame(3);
		MethodBase? methodBase = stackFrame.GetMethod();

		if (methodBase != null && methodBase.DeclaringType != null)
		{
			return methodBase.DeclaringType.Name;
		}

		return "(unknown)";
	}
}
